internal/docstore/dynamodocstore/dynamo.go

```diff
@@ -455,13 +457,22 @@ func (c *collection) put(ctx context.Context, k driver.ActionKind, doc driver.Do
 		in.ConditionExpression = ce.Condition()
 	}
 	_, err = c.db.PutItemWithContext(ctx, in)
-	if err == nil {
-		if newPartitionKey != "" {
-			_ = doc.SetField(c.partitionKey, newPartitionKey)
+	if err != nil {
+		if ae, ok := err.(awserr.Error); ok && ae.Code() == dyn.ErrCodeConditionalCheckFailedException {
+			if k == driver.Create {
+				return gcerr.Newf(gcerr.AlreadyExists, err, "document already exists")
+			}
+			if rev, _ := doc.GetField(docstore.RevisionField); rev == nil && k == driver.Replace {
+				return gcerr.Newf(gcerr.NotFound, nil, "document not found")
+			}
 		}
-		_ = doc.SetField(docstore.RevisionField, rev)
+		return err
 	}
-	return err
+	if newPartitionKey != "" {
+		_ = doc.SetField(c.partitionKey, newPartitionKey)
+	}
+	_ = doc.SetField(docstore.RevisionField, rev)
```

Comment:
nit: just return this line.

Reply:
That would return an error if there were no revision field. We want to support structs that have no revision field.

Reply:
Ah, I see.